{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ab17cd95",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into 'C:/Users/BowheadAI/Documents/R/win-library/4.0'\n",
      "(as 'lib' is unspecified)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  There is a binary version available but the source version is later:\n",
      "             binary  source needs_compilation\n",
      "BiocManager 1.30.12 1.30.15             FALSE\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing the source package 'BiocManager'\n",
      "\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.12 (BiocManager 1.30.15), R 4.0.5 (2021-03-31)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'edgeR'\"\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: C:/Program Files/R/R-4.0.5/library\n",
      "  packages:\n",
      "    boot, class, cluster, KernSmooth, lattice, MASS, Matrix, mgcv, nnet,\n",
      "    spatial, survival\n",
      "\n",
      "Old packages: 'bslib', 'colorspace', 'corrplot', 'curl', 'devtools', 'dplyr',\n",
      "  'ellipsis', 'haven', 'httpuv', 'IRkernel', 'openssl', 'processx',\n",
      "  'reticulate', 'rlang', 'rmarkdown', 'sass', 'statmod', 'stringi', 'tibble',\n",
      "  'tidyselect', 'vctrs'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.12 (BiocManager 1.30.15), R 4.0.5 (2021-03-31)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'Glimma'\"\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: C:/Program Files/R/R-4.0.5/library\n",
      "  packages:\n",
      "    boot, class, cluster, KernSmooth, lattice, MASS, Matrix, mgcv, nnet,\n",
      "    spatial, survival\n",
      "\n",
      "Old packages: 'bslib', 'colorspace', 'corrplot', 'curl', 'devtools', 'dplyr',\n",
      "  'ellipsis', 'haven', 'httpuv', 'IRkernel', 'openssl', 'processx',\n",
      "  'reticulate', 'rlang', 'rmarkdown', 'sass', 'statmod', 'stringi', 'tibble',\n",
      "  'tidyselect', 'vctrs'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.12 (BiocManager 1.30.15), R 4.0.5 (2021-03-31)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'limma'\"\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: C:/Program Files/R/R-4.0.5/library\n",
      "  packages:\n",
      "    boot, class, cluster, KernSmooth, lattice, MASS, Matrix, mgcv, nnet,\n",
      "    spatial, survival\n",
      "\n",
      "Old packages: 'bslib', 'colorspace', 'corrplot', 'curl', 'devtools', 'dplyr',\n",
      "  'ellipsis', 'haven', 'httpuv', 'IRkernel', 'openssl', 'processx',\n",
      "  'reticulate', 'rlang', 'rmarkdown', 'sass', 'statmod', 'stringi', 'tibble',\n",
      "  'tidyselect', 'vctrs'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.12 (BiocManager 1.30.15), R 4.0.5 (2021-03-31)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'org.Hs.eg.db'\"\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: C:/Program Files/R/R-4.0.5/library\n",
      "  packages:\n",
      "    boot, class, cluster, KernSmooth, lattice, MASS, Matrix, mgcv, nnet,\n",
      "    spatial, survival\n",
      "\n",
      "Old packages: 'bslib', 'colorspace', 'corrplot', 'curl', 'devtools', 'dplyr',\n",
      "  'ellipsis', 'haven', 'httpuv', 'IRkernel', 'openssl', 'processx',\n",
      "  'reticulate', 'rlang', 'rmarkdown', 'sass', 'statmod', 'stringi', 'tibble',\n",
      "  'tidyselect', 'vctrs'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"BiocManager\")\n",
    "BiocManager::install(\"edgeR\")\n",
    "BiocManager::install(\"Glimma\")\n",
    "BiocManager::install(\"limma\")\n",
    "BiocManager::install(\"org.Hs.eg.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dda1d687",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: limma\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following object is masked from 'package:limma':\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grDevices':\n",
      "\n",
      "    windows\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in library(\"gplots\"): there is no package called 'gplots'\n",
     "output_type": "error",
     "traceback": [
      "Error in library(\"gplots\"): there is no package called 'gplots'\nTraceback:\n",
      "1. library(\"gplots\")"
     ]
    }
   ],
   "source": [
    "# Load librarys\n",
    "library('edgeR')\n",
    "library('limma')\n",
    "library('Glimma')\n",
    "library('ggplot2')\n",
    "library('RColorBrewer')\n",
    "library('org.Hs.eg.db')\n",
    "library(\"dplyr\")\n",
    "library(\"gplots\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c986a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading principal path L3 counts data to extrac txt files\n",
    "l3_count_main_path <- \"../input/end-als/end-als/transcriptomics-data/L3_counts\"\n",
    "# List of folder in L3_counts\n",
    "l3_count_main_path_list <- list.files(path = l3_count_main_path)\n",
    "# Declare variable just for concatenate txt files\n",
    "all_transcriptomic_counts <- \"Principal\"\n",
    "flag <- TRUE\n",
    "for (folder in l3_count_main_path_list){\n",
    "    # Get sub-folder path\n",
    "    subfolder <- list.files(path=paste(l3_count_main_path, folder, sep='/'))\n",
    "    # If the folder contains more than one file the is excluded\n",
    "    if (length(subfolder) > 1){\n",
    "        next\n",
    "    }\n",
    "    # Read last file this path contain the name of .txt file\n",
    "    exon_txt_file <- list.files(path=paste(l3_count_main_path, folder, subfolder, sep='/'))\n",
    "    # Concatenate all path to read csv\n",
    "    complete_l3_count_path <- paste(l3_count_main_path, folder, subfolder, exon_txt_file, sep='/')\n",
    "    \n",
    "    # Reading exo.txt file\n",
    "    l3_counts_per_id <- read.table(complete_l3_count_path, sep = \"\\t\", header=TRUE)\n",
    "    # Select genes columns and counts\n",
    "    l3_counts_per_id <- subset(l3_counts_per_id, select=(c(1,7)))\n",
    "    # Change the name of columns for genes for genes and counts per id\n",
    "    colnames(l3_counts_per_id) <- c(\"genes\", folder)\n",
    "    # If is the first iterarion then enter into if and save file in all_transcript_count to by concacatenate later\n",
    "    if(flag){\n",
    "        flag<- FALSE\n",
    "        all_transcriptomic_counts <- l3_counts_per_id\n",
    "        next\n",
    "    }\n",
    "    # Join all count in single file\n",
    "    all_transcriptomic_counts <- left_join(x = all_transcriptomic_counts, y = l3_counts_per_id, by = \"genes\", all.x = TRUE)\n",
    "}\n",
    "# Write csv file of raw counts\n",
    "write.csv(all_transcriptomic_counts, \"raw_L3counts_data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bb7ad67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To performance MDS with color by case and ctrl is necesario get the groups from clinical data portal data\n",
    "clinical_data_portal <- read.csv('../input/end-als/end-als/clinical-data/filtered-metadata/metadata/aals_dataportal_datatable.csv')\n",
    "# Creating dataframe just with data that we need\n",
    "\n",
    "\n",
    "clinical_data_portal_subset <- data.frame(clinical_data_portal$Participant_ID, clinical_data_portal$Sex,\n",
    "                                         clinical_data_portal$Race, clinical_data_portal$Subject.Group, \n",
    "                                         clinical_data_portal$Site.of.Onset, clinical_data_portal$NEFH..SMI32.,\n",
    "                                         clinical_data_portal$ISL1, clinical_data_portal$`NKX6.1`, \n",
    "                                         clinical_data_portal$TUBB3..TuJ1., clinical_data_portal$s100b,\n",
    "                                         clinical_data_portal$ALSFRS.R.Baseline, clinical_data_portal$ALSFRS.R.Latest,\n",
    "                                         clinical_data_portal$ALSFRS.R.Progression.Slope,\n",
    "                                         clinical_data_portal$CBS.Baseline, clinical_data_portal$CBS.Latest,\n",
    "                                         clinical_data_portal$CBS.Progression.Slope)\n",
    "# Change col names\n",
    "colnames(clinical_data_portal_subset) <- c(\"Participant_ID\", \"Sex\", \"Race\", \"SubjectGroup\", \"SiteOfOnset\", \"NEFH\",\n",
    "                                          \"ISL1\", \"NKX6\", \"TUBB3\", \"s100b\", \"ALSFRSBaseline\", \"ALSFRSLatest\", \"ALSFRSProgressionSlope\",\n",
    "                                            \"CBSBaseline\", \"CBSLatest\", \"CBS.Progression.Slope\")\n",
    "# Transpose data to join data portal and raw counts data\n",
    "all_transcriptomic_counts_transpose <- t(all_transcriptomic_counts)\n",
    "# Change the columns names for the ids\n",
    "colnames(all_transcriptomic_counts_transpose) <- all_transcriptomic_counts_transpose[1, ]\n",
    "# Delete  ids from the first row \n",
    "all_transcriptomic_counts_transpose <- data.frame(all_transcriptomic_counts_transpose[-c(1), ])\n",
    "# set participant Id as columns\n",
    "all_transcriptomic_counts_transpose$Participant_ID <- rownames(all_transcriptomic_counts_transpose)\n",
    "# Join raw counts and data portal \n",
    "all_transcriptomic_counts_and_data_portal <- left_join(x = all_transcriptomic_counts_transpose, \n",
    "                                       y = clinical_data_portal_subset, by = \"Participant_ID\", all.x = TRUE)\n",
    "\n",
    "# Save transcripts groups to use latter in Difererntial expression analysis\n",
    "transcripts_groups <- all_transcriptomic_counts_and_data_portal[c(\"Participant_ID\", \"Sex\", \"Race\", \n",
    "                                                                 \"SubjectGroup\", \"SiteOfOnset\", \"NEFH\",\n",
    "                                                                 \"ISL1\", \"NKX6\", \"TUBB3\", \"s100b\", \"ALSFRSBaseline\",\n",
    "                                                                 \"ALSFRSLatest\", \"ALSFRSProgressionSlope\", \"CBSBaseline\", \n",
    "                                                                 \"CBSLatest\", \"CBS.Progression.Slope\")]\n",
    "\n",
    "transcripts_groups$SubjectGroup[which(transcripts_groups$SubjectGroup != \"Healthy Control\")]  <- \"ALS\"\n",
    "# Delete columns from raw counts\n",
    "all_transcriptomic_counts_and_data_portal <- subset(all_transcriptomic_counts_and_data_portal, select=-c( Sex, Race, \n",
    "                                                                 SubjectGroup, SiteOfOnset, NEFH,\n",
    "                                                                 ISL1, NKX6, TUBB3, s100b, ALSFRSBaseline,\n",
    "                                                                 ALSFRSLatest, ALSFRSProgressionSlope, CBSBaseline, \n",
    "                                                                 CBSLatest, CBS.Progression.Slope))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13110814",
   "metadata": {},
   "outputs": [],
   "source": [
    "# report process to transpose data\n",
    "\n",
    "# Transpose data to have id in columns\n",
    "all_transcriptomic_counts_and_data_portal <- t(all_transcriptomic_counts_and_data_portal)\n",
    "\n",
    "# Change the columns names for  ids patinets\n",
    "colnames(all_transcriptomic_counts_and_data_portal) <- all_transcriptomic_counts_and_data_portal[60676, ]\n",
    "# Delete  ids from the last row \n",
    "all_transcriptomic_counts_and_data_portal <- all_transcriptomic_counts_and_data_portal[-c(60676), ]\n",
    "# Convert matrix to data frame\n",
    "all_transcriptomic_counts_and_data_portal <- data.frame(all_transcriptomic_counts_and_data_portal)\n",
    "# Set genes ids\n",
    "genes_id <- rownames(all_transcriptomic_counts_and_data_portal)\n",
    "\n",
    "all_transcriptomic_counts_and_data_portal <- as.data.frame(lapply(data.frame(lapply(all_transcriptomic_counts_and_data_portal, as.character), stringsAsFactors=FALSE), as.numeric))\n",
    "#Set row names genes to the new data colletion\n",
    "rownames(all_transcriptomic_counts_and_data_portal) <- genes_id\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "839aac0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# change factor data to char type\n",
    "transcripts_groups$Sex <-  as.character(transcripts_groups$Sex)\n",
    "transcripts_groups$SubjectGroup <-  as.character(transcripts_groups$SubjectGroup)\n",
    "transcripts_groups$SiteOfOnset <-  as.character(transcripts_groups$SiteOfOnset)\n",
    "transcripts_groups$Race <- as.character(transcripts_groups$Race)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9158a86a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter just cases from raw data\n",
    "all_transcriptomic_counts_and_data_portal <- all_transcriptomic_counts_and_data_portal [transcripts_groups$SubjectGroup == 'ALS']\n",
    "transcripts_groups <- transcripts_groups[transcripts_groups$SubjectGroup == 'ALS', ]\n",
    "\n",
    "# Filter by markers\n",
    "# it'is neccesary create boolean list of each marker to filter all missings values\n",
    "# NEFH marker filter\n",
    "NEFH_boolean_list <- !is.na(transcripts_groups$NEFH)\n",
    "all_transcriptomic_counts_and_data_portal <- all_transcriptomic_counts_and_data_portal [NEFH_boolean_list]\n",
    "transcripts_groups <- transcripts_groups[NEFH_boolean_list, ]\n",
    "\n",
    "# ISL1 marker filter\n",
    "ISL1_boolean_list <- !is.na(transcripts_groups$ISL1)\n",
    "all_transcriptomic_counts_and_data_portal <- all_transcriptomic_counts_and_data_portal [ISL1_boolean_list]\n",
    "transcripts_groups <- transcripts_groups[ISL1_boolean_list, ]\n",
    "\n",
    "# NKX6 marker filter\n",
    "NKX6_boolean_list <- !is.na(transcripts_groups$NKX6)\n",
    "all_transcriptomic_counts_and_data_portal <- all_transcriptomic_counts_and_data_portal [NKX6_boolean_list]\n",
    "transcripts_groups <- transcripts_groups[NKX6_boolean_list, ]\n",
    "\n",
    "# TUBB3 marker filter\n",
    "TUBB3_boolean_list <- !is.na(transcripts_groups$TUBB3)\n",
    "all_transcriptomic_counts_and_data_portal <- all_transcriptomic_counts_and_data_portal [TUBB3_boolean_list]\n",
    "transcripts_groups <- transcripts_groups[TUBB3_boolean_list, ]\n",
    "\n",
    "# s100b marker filter\n",
    "s100b_boolean_list <- !is.na(transcripts_groups$s100b)\n",
    "all_transcriptomic_counts_and_data_portal <- all_transcriptomic_counts_and_data_portal [s100b_boolean_list]\n",
    "transcripts_groups <- transcripts_groups[s100b_boolean_list, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c32cc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "dgelist_edgeR_normalize <- DGEList(counts = as.matrix(all_transcriptomic_counts_and_data_portal), group = transcripts_groups$SubjectGroup)\n",
    "# Filter data by cpm\n",
    "keep_data<- filterByExpr(dgelist_edgeR_normalize)\n",
    "dgelist_edgeR_normalize<- dgelist_edgeR_normalize[keep_data, , keep.lib.sizes=FALSE]\n",
    "# Normalize Data\n",
    "edgeR_normalize_data <- calcNormFactors(dgelist_edgeR_normalize)\n",
    "dgelist_edgeR_normalize <- DGEList(counts = as.matrix(edgeR_normalize_data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "751f1075",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot MDS\n",
    "plotMDS(dgelist_edgeR_normalize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba70ecef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate count per million and keep samples\n",
    "cpm_quantile_transform <- cpm(all_transcriptomic_counts_and_data_portal)\n",
    "# thresh_cpm to filter data\n",
    "thresh_cpm <- cpm_quantile_transform > 0.5\n",
    "keep_quantile_data <- rowSums(thresh_cpm) >= 2\n",
    "counts.keep_quantile_data <- all_transcriptomic_counts_and_data_portal[keep_quantile_data,]\n",
    "# # Quantile normalization\n",
    "quantile_normalization_data <- normalizeQuantiles(counts.keep_quantile_data, ties=TRUE)\n",
    "# # log transformation\n",
    "log_quantile_norm_data <- log1p(quantile_normalization_data)\n",
    "# #created dglist to make MDS\n",
    "dgelist_quantile_data <- DGEList(counts = as.matrix(log_quantile_norm_data), group = transcripts_groups$SubjectGroup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33f4c142",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot MDS\n",
    "plotMDS(dgelist_quantile_data, col=as.numeric(dgelist_quantile_data$samples$group))\n",
    "legend(\"bottomleft\", as.character(unique(dgelist_quantile_data$samples$group)), col=1:3, pch=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51bea4ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove bath effect by markers\n",
    "quantile_data_removeBatchEffect <- removeBatchEffect(log_quantile_norm_data, covariates=transcripts_groups[c(\"NEFH\", \"ISL1\", \"NKX6\", \"TUBB3\", \"s100b\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ab939be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove bath effect by markers\n",
    "quantile_data_removeBatchEffect <- removeBatchEffect(log_quantile_norm_data, covariates=transcripts_groups[c(\"NEFH\", \"ISL1\", \"NKX6\", \"TUBB3\", \"s100b\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e29f484e",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_corrected_data <- heatmap.2(quantile_data_removeBatchEffect, dendrogram='both', distfun = function(x) dist(x, method=\"euclidean\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f0729a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "hc <- as.hclust( heatmap_corrected_data$colDendrogram )\n",
    "clusters <- cutree(hc, k=2 )\n",
    "# clusters_corrected_data <\n",
    "transcripts_groups$clusters <- clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65dbb7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#statistical comparation\n",
    "lapply(transcripts_groups[,6:10], function(x) kruskal.test(x ~ clusters, data=transcripts_groups))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fd9ce78",
   "metadata": {},
   "outputs": [],
   "source": [
    "lapply(transcripts_groups[,11:16], function(x) kruskal.test(x ~ clusters, data=transcripts_groups))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f06abe35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diferential expression analisis\n",
    "# Declare covariantes\n",
    "NKX <- transcripts_groups$NEFH\n",
    "NEFH <- transcripts_groups$ISL1\n",
    "TUBB3 <- transcripts_groups$NKX6\n",
    "ISL1 <- transcripts_groups$TUBB3\n",
    "s100b <- transcripts_groups$s100b\n",
    "sex <- transcripts_groups$Sex\n",
    "race <- transcripts_groups$Race\n",
    "clusters <- factor(transcripts_groups$clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0622e53",
   "metadata": {},
   "outputs": [],
   "source": [
    "design_matrix <- model.matrix(~0 + clusters + NKX + NEFH + TUBB3 + ISL1 + s100b + race + sex )\n",
    "\n",
    "colnames(design_matrix) <- c('clusters_one1', 'clusters_one2','NKX', 'NEFH', 'TUBB3', 'ISL1', 's100b','race1', 'race2','race3', 'rece4', 'sex') \n",
    "\n",
    "# fit data\n",
    "fit_case_clusters <- lmFit(as.matrix(log_quantile_norm_data), design_matrix)\n",
    "# contrast case vs control data\n",
    "cont.matrix_cluster <- makeContrasts(clusters_one1-clusters_one2, levels = design_matrix)\n",
    "fit_case_clusters <- contrasts.fit(fit_case_clusters, cont.matrix_cluster)\n",
    "fit_case_clusters <- eBayes(fit_case_clusters)\n",
    "\n",
    "# cluster1 vs all\n",
    "DEP_SIG_ALS_cluster1_vs_cluster2 <- topTable(fit_case_clusters, adjust = \"BH\", coef = 1, number = length(fit_case_clusters$coefficients)) \n",
    "# Filter by  p.adjust.val\n",
    "DEP_SIG_ALS_cluster1_vs_cluster2 <- DEP_SIG_ALS_cluster1_vs_cluster2[ DEP_SIG_ALS_cluster1_vs_cluster2$adj.P.Val < 0.05, ]\n",
    "DEP_SIG_ALS_cluster1_vs_cluster2\n",
    "write.csv(DEP_SIG_ALS_cluster1_vs_cluster2, 'cluster1_vs_cluster2_corrected.csv') # Cluster"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
